pkg/controller/nodetune/sync_jobs.go

```diff
@@ -138,54 +145,113 @@ func (ncdc *Controller) makePerftuneJobForContainers(ctx context.Context, podSpe
 	)
 }
 
-func (ncdc *Controller) makeJobForContainers(ctx context.Context) (*batchv1.Job, error) {
+func (ncdc *Controller) makeResourceLimitJobsForContainers(ctx context.Context, podSpec *corev1.PodSpec, scyllaPods []*corev1.Pod, scyllaContainerIDs []string) (*batchv1.Job, error) {
+	scyllaProcessHostPIDs := make([]int, 0, len(scyllaPods))
+
+	for _, scyllaPod := range scyllaPods {
+		containerID, err := getScyllaContainerIDInCRIFormat(scyllaPod)
+		if err != nil {
+			return nil, fmt.Errorf("can't get scylla containerID in CRI format: %w", err)
+		}
+		containerStatus, err := ncdc.criClient.Inspect(ctx, containerID)
+		if err != nil {
+			return nil, fmt.Errorf("can't inspect Scylla Pod %q container %q: %w", naming.ObjRef(scyllaPod), containerID, err)
+		}
+
+		if containerStatus.Info.PID == nil {
+			continue
+		}
+
+		scyllaProcessHostPIDs = append(scyllaProcessHostPIDs, *containerStatus.Info.PID)
+	}
+
+	cr, err := ncdc.newOwningDSControllerRef()
+	if err != nil {
+		return nil, fmt.Errorf("can't get controller ref: %w", err)
+	}
+
+	if len(scyllaProcessHostPIDs) == 0 {
+		klog.Errorf("can't determine Scylla container process PIDs from CRI, their resource process limits won't be raised")
+		return nil, nil
+	}
+
+	rlimitsJobs, err := makeRlimitsJobForContainers(cr, ncdc.namespace, ncdc.nodeConfigName, ncdc.nodeName, ncdc.nodeUID, ncdc.operatorImage, podSpec, scyllaContainerIDs, scyllaProcessHostPIDs)
+	if err != nil {
+		return nil, fmt.Errorf("can't make rlimits jobs: %w", err)
+	}
+
+	return rlimitsJobs, nil
+}
+
+func (ncdc *Controller) makeJobsForContainers(ctx context.Context) ([]*batchv1.Job, error) {
 	localScyllaPods, err := ncdc.localScyllaPodsLister.List(naming.ScyllaSelector())
 	if err != nil {
 		return nil, fmt.Errorf("can't list local scylla pods: %w", err)
 	}
 
-	var optimizablePods []*corev1.Pod
-	var scyllaContainerIDs []string
+	localScyllaPods = slices.FilterOut(localScyllaPods, func(pod *corev1.Pod) bool {
+		return pod.DeletionTimestamp != nil
+	})
 
-	for i := range localScyllaPods {
-		scyllaPod := localScyllaPods[i]
+	// Sort Pods to have stable representation for the same set of Pods.
+	sort.Slice(localScyllaPods, func(i, j int) bool {
+		return localScyllaPods[i].Name < localScyllaPods[j].Name
+	})
```

Comment:
It's not good enough imo because it requires diving into multiple subsequent calls to figure out why it's there in the first place. This comment doesn't explain why it's needed but just says what is done a line below. If you're worried about the comment getting out of date, it can be made clearer by just sorting the very thing that needs it, otherwise the next person to work on it will have to go through every possible side effect that sorting the entire slice of pods could introduce.

Reply:
Sorting objects is common pattern in our codebase, probably all sort usages refer to listers/informers returning random order objects which are later projected into API object. Without sorting it would result in endless updates. Here we are doing it for the same purpose and comment explains that. We want the same representation (of containerIDs, names, PIDs - anything) for the same set of Pods.

>  If you're worried about the comment getting out of date, it can be made clearer by just sorting the very thing that needs it

Should I sort it right before Jobs are defined? That would be multiple places and I'm against it.

> otherwise the next person to work on it will have to go through every possible side effect that sorting the entire slice of pods could introduce.

Would you mind pointing out what are those side effects?

Reply:
> If it's there, it's because it's required.

Just because *a* caller does something doesn't imply a callee requires that. 

`makePerftuneJobForContainers` takes `optimizablePods` and `optimizableScyllaContainerIDs`, both of which would be sorted by pods' names. Are both slices expected to be sorted? Two comments above you said it's just the container IDs.

> Sorting objects is common pattern in our codebase, probably all sort usages refer to listers/informers returning random order objects which are later projected into API object. Without sorting it would result in endless updates. 

It is, but it's much closer to where it's being used and so it's clear from the near context why it's required. It's not the case here.

> That would be multiple places and I'm against it.

Why?

> Would you mind pointing out what are those side effects?

See the first point. If the answer is no, then you have your example.

Reply:
> Are both slices expected to be sorted? Two comments above you said it's just the container IDs.

Yes they are. I haven't said it's just for container IDs, I meant set of values that we extract from Pods may change tomorrow.

 It is, but it's much closer to where it's being used and so it's clear from the near context why it's required. It's not the case here.




> It is, but it's much closer to where it's being used and so it's clear from the near context why it's required. It's not the case here.
> 
> > That would be multiple places and I'm against it.
> 
> Why?

Because it's not efficient and unnecessary. Pod sorting was already there, I moved it upper in the call chain because there's a second function requiring this order and it can be done once.  

> > Would you mind pointing out what are those side effects?
> 
> See the first point. If the answer is no, then you have your example.

I don't understand. "No" to "why" question doesn't make any sense.

Reply:
> > > If it's there, it's because it's required.
> >
> > Just because a caller does something doesn't imply a callee requires that.
> > Are both slices expected to be sorted? Two comments above you said it's just the container IDs.
> 
> Yes they are. 

How? `optimizablePods` weren't sorted before, only the container IDs were, and you didn't modify any related code in `makePerftuneJobForContainers`. So what changed to make the order in `optimizablePods` required?

> Because it's not efficient and unnecessary.

What's the difference in efficiency? Sorting the slice twice vs once? A small price to pay for improved readability in my opinion.

> Pod sorting was already there,

Where? I only see sorting of containerIDs being removed in your commits.

> > See the first point. If the answer is no, then you have your example.
> >
> I don't understand. "No" to "why" question doesn't make any sense.

Huh? `Are both slices expected to be sorted?`.

Reply:
> How? optimizablePods weren't sorted before, only the container IDs were, and you didn't modify any related code in makePerftuneJobForContainers. So what changed to make the order in optimizablePods required?

containerIDs are still sorted. What's consumed from optimizablePods is required to be sorted (list of data directories) and is unfortunately sorted second time due to how map traversal works. Yet it's not worse than before.

> What's the difference in efficiency? Sorting the slice twice vs once? 

Efficiency is the same, instead sorting containerIDs, we sort Pods resulting in sorted containerIDs and having sorted Pods for other user. While your proposal is to sort twice. Given that this controller runs under very low resources, efficiency is more important.

> A small price to pay for improved readability in my opinion.

Technical facts comes before personal opinions.

> Huh? Are both slices expected to be sorted?.

Please don't leave riddles in your comments, be specific what you mean, otherwise it's not moving us forward.

Reply:
> Why does object sorting matters for the updates? We sort API array fields to make sure they remain idempotent to changes and that's probably better done anyways as a way of defensive coding and making sure things don't change through say a random map iteration, possibly added later in some helper etc. I am not sure why Jobs themselves would need to be sorted.

I think you misunderstood what is being sorted here. I'm not sorting Jobs, but Pods. Having consistent order of Pods results in sorted fields of those which later are inputs to Jobs spec where order is important to avoid endless updates.

Reply:
> > Sorting objects is common pattern in our codebase
> 
> I've got just 1 hit for that "pattern" among all the functions that sort/return runtime.Object ` grep -RIni ./pkg/controller/ -e 'sort\.'` https://github.com/scylladb/scylla-operator/blame/6e96cd1d97028a95cf34ff67a846155ae5972c26/pkg/controller/scylladbdatacenter/resource.go#L1407
> 
> Everywhere else I've peaked, we sort the API array fields.

why it has to implement runtime.Object to be considered as a pattern?

Reply:
> What's consumed from optimizablePods is required to be sorted (list of data directories) and is unfortunately sorted second time due to how map traversal works.

So `optimizablePods` don't need to be sorted. So `If it's there, it's because it's required` doesn't really hold as an argument against sorting it prematurely and not explaining why, does it?

> Technical facts comes before personal opinions.

The difference is literally a constant. And my "personal opinion" refers to maintainability and readability of the code. This is not the self-documenting code that you think it is. And I first asked you only to leave an explanation for sorting the entire slice of pods so far ahead of where parts of it are consumed.

Let me reiterate - imagine you're new to the codebase. If you needed to change something in that piece of code, to figure out why the entire slice of pods is being sorted, you have to go into the implementation of the called functions that consume anything that spanned from it to backtrack to which parts needed sorting any why. This is the price you pay for doing it this way. As well as introducing potential unwanted dependencies on the order which doesn't need to be there. It's not just a style preference.

Reply:
> So optimizablePods don't need to be sorted by themselves. So If it's there, it's because it's required doesn't really hold as an argument against sorting it prematurely and not explaining why, does it?

Then remove the sorting and observe what happens - exactly what comment explains - inconsistent representation for the same set of Pods. Add it back again and now you've got consistent representation for the same set of Pods - exactly what comment why sorting is there. 

> And my "personal opinion" refers to maintainability and readability of the code. This is not the self-documenting code that you think it is.

I have a different opinion, and I don't agree that your suggestion would help with making it more self explanatory.

> And I first asked you only to leave an explanation for sorting the entire slice of pods so far ahead of where parts of it are consumed.

It's already explained why, I'm not sure why it's being ignored. 

> Let me reiterate - imagine you're new to the codebase. If you needed to change something in that piece of code, to figure out why the entire slice of pods is being sorted, you have to go into the implementation of the called functions that consume anything that spanned from it to backtrack to which parts needed sorting any why. This is the price you pay for doing it this way. 

Can't you just accept the fact that pods are sorted and move on? Unless you manipulate it, it shouldn't be your concern. You either take benefit of it, or you don't have to care.

> As well as introducing potential unwanted dependencies on the order which doesn't need to be there. It's not just a style preference.

You call it potential unwanted dependencies, I call it preprocessing. When you split function into multiple smaller ones these inherit properties of data structures passed along. 
The same way we don't always validate if passed pointers are non-nil. Isn't it unwanted dependency on object availability potentially leading to crashing?

Reply:
> I think you misunderstood what is being sorted here. I'm not sorting Jobs, but Pods. Having consistent order of Pods results in sorted fields of those which later are inputs to Jobs spec where order is important to avoid endless updates.

My main point was to say, it's better/safer to sort it where matters, not in the inputs higher in the chain. No other caller can break this and the reason for sorting is then obvious and localized.

For this case I assume (it's not cleae from the sort or the comment) that it's about
`localScyllaPods->scyllaPod->containerID->optimizableScyllaContainerIDs->ncdc.makePerftuneJobForContainers()`
in which case I'd expect the safe sort around here
https://github.com/scylladb/scylla-operator/blob/69ff1676dfdd4796512278c877e8891aaaa38886/pkg/controller/nodetune/resource.go#L208

Reply:
Updated the comment after offline discussion.

Reply:
As discussed on the call, the comment is good enough for me. @tnozicka does that also work for you? 
(I'll give it a lgtm if Tomas doesn't reply by tomorrow since he already gave it an approval)